### PR TITLE
fix(ext.commands): required is False although default is None

### PR DESCRIPTION
**Pull request by Krutyi-4el**
## Summary

https://github.com/Pycord-Development/pycord/issues/2089#issuecomment-1836791234

## Information

- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [ ] I have updated the changelog to include these changes.

### CODE REVIEW

**Comment by Dorukyum** [comment]
I don't understand, there's no way `x or y` returns `False` when one side is `True`

**Comment by Lulalaby** [comment]
Yeah same thought here. Closing it as invalid.

**Comment by Krutyi-4el** [comment]
@Lulalaby @Dorukyum is this code valid?

```python
@bot.bridge_command()
async def test_no_default(ctx, text: Option(str, required=False)):
    await ctx.respond(text or "hello")```

I think it should work but it doesn't because the default is None.

**Comment by plun1331** [comment]
> I don't understand, there's no way `x or y` returns `False` when one side is `True`

The issue seems to be that it's returning `True` when it logically shouldn't (`required` is `False`, however `default` is `None`)

`default` being `None` should not change the behavior of `required`

**Comment by Dorukyum** [comment]
> `default` being `None` should not change the behavior of `required`

Isn't that how it's always been? `Option(...)` would be required while `Option(..., default=None)` would be optional.

**Comment by Krutyi-4el** [comment]
> > `default` being `None` should not change the behavior of `required`
> 
> Isn't that how it's always been? `Option(...)` would be required while `Option(..., default=None)` would be optional.

Yes, but it didn't (and still doesn't) work for bridge commands invoked via prefix.

This PR fixes the issue.

**Comment by plun1331** [comment]
> > `default` being `None` should not change the behavior of `required`
> 
> Isn't that how it's always been? `Option(...)` would be required while `Option(..., default=None)` would be optional.

But if required is `False` with default being `None`, it automatically flips `required` to `True`

**Comment by Krutyi-4el** [comment]
> But if required is `False` with default being `None`, it automatically flips `required` to `True` (at least, that's my understanding of this pr)

Yes. Because of this the command above doesn't work without the argument.
The most confusing part is that it works when invoked as a slash command.

By the way, looks like `Option` alrea

**Comment by Krutyi-4el** [comment]
Missclick 😅

**Comment by Dorukyum** [comment]
> By the way, looks like `Option` already sets its `.required` attribute properly. Maybe it should be just `required = param.annotation.required`.

My bad, I understand the pr now and I think you're right. We can't do `and default is None` because `default` doesn't have to be `None`. Like you said, `Option` already sets `required` meaning we can just use `required = param.annotation.required` here.

File: discord/ext/commands/core.py [hunk 1]
```diff
@@ -565,7 +565,7 @@ async def dispatch_error(self, ctx: Context, error: Exception) -> None:
     async def transform(self, ctx: Context, param: inspect.Parameter) -> Any:
         if isinstance(param.annotation, Option):
             default = param.annotation.default
-            required = param.annotation.required or default is None
+            required = param.annotation.required and default is None
```
Comment (Dorukyum) [inline discord/ext/commands/core.py:568]:
```suggestion
            required = param.annotation.required
```

**Review by Dorukyum — CHANGES_REQUESTED** [review]

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**